### PR TITLE
Prevent duplicate assertLoc requests from showing up

### PR DESCRIPTION
**Pull request by vicmgs**

### CODE REVIEW

File: lib/blockchain_node/watcher.ex [hunk 1]
```diff
@@ -171,4 +159,27 @@ defmodule BlockchainNode.Watcher do
       time: current_time
     }
   end
+  defp add_assert_location_token(txn, type, ownerAddress, gatewayAddress, location) do
+    token = :crypto.strong_rand_bytes(32)
```
Comment (allenan) [inline lib/blockchain_node/watcher.ex]: could you use the h3 index as the token in this case?

File: lib/blockchain_node/watcher.ex [hunk 2]
```diff
@@ -94,40 +94,28 @@ defmodule BlockchainNode.Watcher do
...
-        })
-      )
-    end)
+    gatewayAddress = to_string(:libp2p_crypto.address_to_b58(:blockchain_txn_assert_location_v1.gateway_address(txn)))
+    ownerAddress = to_string(:libp2p_crypto.address_to_b58(:blockchain_txn_assert_location_v1.owner_address(txn)))
+    location = to_string(:h3.to_string(:blockchain_txn_assert_location_v1.location(txn)))
+    type = "assertLocationRequest"
+
+    existing_tokens = Gateways.get(:tokens)
+    duplicate_token =
```
Comment (allenan) [inline lib/blockchain_node/watcher.ex]: isn't the condition for it to be duplicate that the location is the same? i don't see that being checked here

File: lib/blockchain_node/watcher.ex (hunk 1 above)
Comment (vicmgs) [reply lib/blockchain_node/watcher.ex]: yea that works, is it bad to generate a random token?

Comment (allenan) [reply lib/blockchain_node/watcher.ex]: I was just trying to think of a way to do the dedup logic without scanning all the existing requests. i guess you'd need to use some combination of gateway address + location, but if you used that for the token, and ignore requests where a token exists, then maybe that'd work?

Comment (vicmgs) [reply lib/blockchain_node/watcher.ex]: with the dedup in place, there wouldn't be many tokens at all though

File: lib/blockchain_node/watcher.ex (hunk 2 above)
Comment (allenan) [reply lib/blockchain_node/watcher.ex]: also from the CH ticket: `If there's some contention (different locations) I'd argue it should show the most precise request (ie. the one with the most digits in the H3 index)`

Comment (vicmgs) [reply lib/blockchain_node/watcher.ex]: its under the case duplicate_token ==> if clause

Comment (allenan) [reply lib/blockchain_node/watcher.ex]: oh yeah i see it down there. it's just a bit of a complicated code flow i suppose. which is why i was trying to think of simpler ways of checking for duplicates

File: lib/blockchain_node/watcher.ex [hunk 3]
```diff
@@ -94,40 +94,28 @@ defmodule BlockchainNode.Watcher do
...
+        gatewayAddress == to_string(:libp2p_crypto.address_to_b58(:blockchain_txn_assert_location_v1.gateway_address(t.txn))) and
+        ownerAddress == to_string(:libp2p_crypto.address_to_b58(:blockchain_txn_assert_location_v1.owner_address(t.txn))) and
+        type == "assertLocationRequest"
+      end)
+
+    case duplicate_token do
+      nil ->
+        add_assert_location_token(txn, type, ownerAddress, gatewayAddress, location)
+      _ ->
+        if location != to_string(:h3.to_string(:blockchain_txn_assert_location_v1.location(duplicate_token.txn))) do
```
Comment (allenan) [inline lib/blockchain_node/watcher.ex]: we should centralize this messy to_string logic, I think the same is used in Gateways as well

File: lib/blockchain_node/gateways/gateways.ex [hunk 4]
```diff
@@ -225,7 +226,7 @@ defmodule BlockchainNode.Gateways do
   end
 
   defp geo_to_h3({_lat, _lng} = coordinates, resolution) do
-    :h3.from_geo(coordinates, resolution) |> :h3.to_string() |> to_string()
+    :h3.from_geo(coordinates, resolution) |> Helpers.to_h3_string()
```
Comment (allenan) [inline lib/blockchain_node/gateways/gateways.ex]: could we combine from_geo into the helper too? i think we'd always call these 2 together

Comment (vicmgs) [reply lib/blockchain_node/gateways/gateways.ex]: hmmm not in watcher, watcher is :blockchain_txn_assert_location_v1.location() |> Helpers.to_h3_string

Comment (allenan) [reply lib/blockchain_node/gateways/gateways.ex]: ok

File: lib/blockchain_node/watcher.ex [hunk 5]
```diff
@@ -94,40 +94,42 @@ defmodule BlockchainNode.Watcher do
...
+    case old_token do
+      nil ->
+        add_assert_location_token(txn, type, ownerAddress, gatewayAddress, location)
+      _ ->
+        old_location =
+          old_token.txn
+          |> :blockchain_txn_assert_location_v1.location()
+          |> Helpers.to_h3_string()
+
+        if location != old_location && String.length(location) >= String.length(old_location) do
```
Comment (allenan) [inline lib/blockchain_node/watcher.ex]: does string length always determine a more precise location?

Comment (vicmgs) [reply lib/blockchain_node/watcher.ex]: double checking

**Pull request merged**